rclpy/src/rclpy/_rclpy_signal_handler.c

```diff
@@ -21,92 +21,182 @@
 #else
-  #define SIGNAL_HANDLER_T sig_t
+#define SIGNAL_HANDLER_T sig_t
 #endif  // _WIN32
 
-/// Global reference to original signal handler for chaining purposes
-SIGNAL_HANDLER_T g_original_signal_handler = NULL;
+#define NULL_SIGNAL_HANDLER NULL
 
-typedef _Atomic (rcl_guard_condition_t **) atomic_rcl_guard_condition_ptrptr_t;
+#define DEFINE_SIGNAL_HANDLER(name) \
+  static void _ ## name(int signum); \
+  static SIGNAL_HANDLER_T get_ ## name() {return _ ## name;} \
+  static bool is_ ## name(SIGNAL_HANDLER_T handler) {return _ ## name == handler;} \
+  static void _ ## name(int signum)
 
-/// Global reference to guard conditions
-/// End with sentinel value instead of count to avoid mismatch if signal
-/// interrupts while adding or removing from the list
-atomic_rcl_guard_condition_ptrptr_t g_guard_conditions;
+#define SIGNAL_HANDLER_ARGS signum
 
-/// Warn if getting g_guard_conditions could deadlock the signal handler
-/// \return 0 if no exception is raised, -1 if an exception was raised
-static int
-check_signal_safety()
+#define call_signal_handler(handler, ...) handler(__VA_ARGS__)
+
+#define install_signal_handler(signum, handler) signal(signum, handler)
+
+#define is_null_signal_handler(signal_handler) \
+  (NULL_SIGNAL_HANDLER == signal_handler)
+
+#endif
+
+// Forward declarations
+static bool trigger_guard_conditions();
+static void unregister_sigint_signal_handler();
+
+/// Original signal handler for chaining purposes
+SIGNAL_HANDLER_T g_original_sigint_handler = NULL_SIGNAL_HANDLER;
+
+/// Signal handler function
+DEFINE_SIGNAL_HANDLER(rclpy_sigint_handler)
 {
-  static bool did_warn = false;
-  if (!did_warn && !atomic_is_lock_free(&g_guard_conditions)) {
-    did_warn = true;
-    const char * deadlock_msg =
-      "Global guard condition list access is not lock-free on this platform."
-      "The program may deadlock when receiving SIGINT.";
-    return PyErr_WarnEx(PyExc_ResourceWarning, deadlock_msg, 1);
+  if (!is_null_signal_handler(g_original_sigint_handler)) {
+    call_signal_handler(g_original_sigint_handler, SIGNAL_HANDLER_ARGS);
   }
-  return 0;
-}
 
-// Forward declaration
-static void catch_function(int signo);
+  if (!trigger_guard_conditions()) {
+    // There may have been another signal handler chaining to this
+    // one when we last tried to unregister ourselves.
+
+    // Try to unregister again.
+    unregister_sigint_signal_handler();
+  }
+}
 
 /// Restore the original signal handler when ours was registered
 static void
-restore_original_signal_handler()
+unregister_sigint_signal_handler()
 {
-  const SIGNAL_HANDLER_T current_handler = signal(SIGINT, g_original_signal_handler);
-  if (current_handler != catch_function) {
-    // Oops, someone else must have registered a signal handler that chains to us
-    // put it back so it continues to work
-    signal(SIGINT, current_handler);
+  if (is_null_signal_handler(g_original_sigint_handler)) {
+    // Handler not registered yet
+    return;
```

Comment:
I guess we know the cpython signal handler should be installed before rclpy's, but it is possible something else uninstalled the original signal handler before rclpy's was installed. In that case `g_original_sigint_handler` is `NULL` and the current signal handler is rclpy's so `return` will leave rclpy's handler installed indefinitely.

Reply:
Good catch, fixed in 2862a945af8237f3b384dfb7fb88d9fc4cf8a090.